### ISSUE TITLE
Allow side-channel P2P messaging between nodes

### ISSUE THREAD

**Issue by srmo** [tendermint/tendermint#476]
We would like to have something like a side-channel for arbitrary messaging between nodes. Such messages should not influence consensus/mempool. One solution is to write a custom p2p layer on top of our ABCI app(s) but why not use the existing p2p network established by the nodes?
We could have a subscribe endpoint (like provided for WSRPC)
Could you implement such a thing? If yes, how difficult would that be?

something like `/broadcast_sidechannel?msg=“Hi There!”`

**Comment by adrianbrink** [comment]
@srmo What is the use case for this feature?

**Comment by srmo** [comment]
@adrianbrink 
We require some kind of toplevel protocol to determine if certain blockchain-persisted transactions should be displayed to clients (client applications). This protocol requires a fully anonymized broadcast exchange of multiple message types. Using the blockchain to run the protocol wastes time and storage. It seems obvious that this should be handled off-chain. The current approach is writing a second P2P between all the chain clients. But why write a new one when we could use the already existing p2p transport layer.

Does that make it clearer or do you need more information?

**Comment by adrianbrink** [comment]
@srmo I don't think that we will implement this in tendermint, as we are focusing in getting tendermint ready for COSMOS. This seems to be a very specific use case that isn't needed for COSMOS and hence I will be closing this. 

@ebuchman Maybe it could be an idea to discuss how to enable users of tendermint to write their own extensions.

**Comment by srmo** [comment]
sadface. @adrianbrink I'm looking forward to a COSMOS road map. We are using tendermint for non-cosmos stuff so we are very interested in seeing what can be expected in terms of new functionality.

**Comment by ebuchman** [comment]
I wouldn't close this so quickly :)

**Comment by rigelrozanski** [comment]
Interesting feature. I bet there is a use case for this communication layer in implementing something like proxy nodes for validators.

**Comment by srmo** [comment]
Any new thoughts on this?

**Comment by paralin** [comment]
If libp2p was in use, then the "side-channel" would just be libp2p pub-sub / encrypted channels between peers.

**Comment by ebuchman** [comment]
It's easy enough to add another Reactor for the tendermint p2p, as was done here: https://github.com/tendermint/tendermint/pull/761

We're interested in this, but it's not priority for launch and we don't really have the time to incorporate it yet. But perhaps we can merge it soon and keep it behind a feature flag so folks can optionally enable if they want for their private networks.

**Comment by Ivshti** [comment]
Yeah, this feature is very cool, but is it really in the scope of tendermint? 
Especially as @paralin says, if it uses libp2p then the underlying swarm can be re-used/accessed and messages can just be sent arbitrarily

**Comment by srmo** [comment]
Interesting. And yes, if this is exposed via RPC this would be amazing!

**Comment by srmo** [comment]
@Ivshti @paralin what does the scenario/approach look like with libp2p? Are you suggesting using libp2p in tendermint and then any external app could facilitate that? I'm not very well versed in those aspects.

**Comment by paralin** [comment]
There are a few projects (mine included) that seek to make a mesh network of "agents" over arbitrary transports. Tendermint and other projects could basically piggy-back on that framework to do communication and not have to worry about the underlying security / encryption / routing /etc.

**Comment by srmo** [comment]
Can we have any news on this? Especially as there is work going on in the broadcast section of transactions. Any ideas? ADR required? What about the other suggestions, raised in this thread?

**Comment by ebuchman** [comment]
Unfortunately, despite it being over a year later, the position on this still hasn't really changed. 

We're increasingly interested in moving to libp2p, which is something we hope to achieve this year. Hopefully that will ease the friction for users seeking to add additional protocols without requiring changes to Tendermint. 

One thing we could consider sooner is slight refactors to the `node/node.go` to support users injecting custom reactors onto the switch. That way users could maintain their own `tendermint-extended` without needing to fork tendermint itself, where the only code in `tendermint-extended` would be a simple [main.go file](https://github.com/tendermint/tendermint/blob/develop/cmd/tendermint/main.go) plus whatever custom reactors/logic they want. 

Such a change to `node.go` would probably help / bear on issues like #2237, #2233, #1686, #1148.

Sorry if this isn't the answer you wanted to hear, but we have to be careful with adding orthogonal features to the codebase that might be better incorporated through future architecture changes.

**Comment by srmo** [comment]
Thanks! I totally hear you and the team, regarding such major features. That's why your first-aid idea sounds really good! Is there already something in the making regarding `node.go` changes?

**Comment by ebuchman** [comment]
There is nothing in the making afaik. Looking at it now though, it might be a bit unwieldily to factor out. That said, the `Switch()` method on the `Node` gives access to the underlying p2p switch, which means you should already be able to add your own reactors without much trouble.

**Comment by ebuchman** [comment]
Alright if I close this for https://github.com/tendermint/tendermint/issues/3176 for now and we can push the experimentation into external repos?

**Comment by srmo** [comment]
Alright. Let’s do this. I’m keeping an eye on this feature :)

**Comment by akv0** [comment]
Hi.
Just for sharing our experience, we are using Tendermint and libp2p as side channel for extra messaging. Our GoLang package is named `subpub` and could be easily integrated into any GoLang Tendermint based blockchain

https://gitlab.com/vocdoni/go-dvote/-/tree/master/subpub
